### PR TITLE
Integration test #57

### LINKED ISSUE #57: Add integration tests for basic CRUD operations
We can add integration tests using `integration_test` package. Refer this [docs](https://docs.flutter.dev/cookbook/testing/integration/introduction) for more details.

We can add integration tests around CRUD operations for notes. Initially we can use `Continue as guest` as option in signup page, create a new note with title and body, save it and check it is displayed in home page. 

We can similarly check by updating an existing note to check if updation is working, and delete a note and check if it is removed from home page

### PR DESCRIPTION
**Pull request by PratyushChauhan**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [x] Testing

## Description
Add integration tests
## Featured Covered in this PR

- [x] Guest login integration tests.
- [x] Make note integration test.
- [x] Make note deletion integration test.
- [x] Make note updation integration test.

## Related Tickets & Documents

- Related Issue #
- Closes #57 

## Screenshots, Recordings

*Please replace this line with instructions on how to test your changes, or any screenshots or recording you can attach here.*

## Tested Feature??

- [x] In Real Device.
- [ ] In Emulator

### CODE REVIEW

File: integration_test/test_guest_login.dart [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import 'package:dairy_app/app/view/app.dart';
+import 'package:dairy_app/core/pages/home_page.dart';
+import 'package:dairy_app/features/auth/presentation/pages/auth_page.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:integration_test/integration_test.dart';
+
+Future<void> main() async{
+  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
+
```
Comment (SankethBK) [inline integration_test/test_guest_login.dart:9]: @PratyushChauhan can you try calling `di.init()` as well here
```
import 'core/dependency_injection/injection_container.dart' as di;

...
  await di.init();

```

Comment (PratyushChauhan) [reply integration_test/test_guest_login.dart:9]: tried this, got an async gap error.
```
PS C:\dev\projects\OSS\diaryvault> flutter test integration_test/test_guest_login.dart                                                                                                                                                                                 
00:00 +0: loading C:/dev/projects/OSS/diaryvault/integration_test/test_guest_login.dart                                                                                                                                                                                             R02:00 +0: loading C:/dev/projects/OSS/diaryvault/integration_test/test_guest_login.dart                                                                                                                                                                                       118.6s
√  Built build\app\outputs\flutter-apk\app-debug.apk.
02:04 +0: loading C:/dev/projects/OSS/diaryvault/integration_test/test_guest_login.dart                                                                                                                                                                                             I02:12 +0: loading C:/dev/projects/OSS/diaryvault/integration_test/test_guest_login.dart                                                                                                                                                                                         7.9s
02:17 +0: loading C:/dev/projects/OSS/diaryvault/integration_test/test_guest_login.dart                                                                                                                                                                                             
🐛  AuthSessionBloc: AuthSessionBloc recreated
💡  KeyValueDataSource: searching for key current_theme
02:17 +0 -1: loading C:/dev/projects/OSS/diaryvault/integration_test/test_guest_login.dart [E]                                                                                                                                                                                      
  Failed to load "C:/dev/projects/OSS/diaryvault/integration_test/test_guest_login.dart": Instance of 'NotInitializedError'
  package:flutter_dotenv/src/dotenv.dart 41:7                                                                 DotEnv.env
  package:dairy_app/features/sync/data/datasources/dropbox_sync_client.dart 23:39                             new DropboxSyncClient
  package:dairy_app/core/dependency_injection/injection_container.dart 135:7                                  init
  ===== asynchronous gap ===========================
  c:/dev/projects/OSS/diaryvault/integration_test/test_guest_login.dart 10:3                                  main
  ===== asynchronous gap ===========================
  c:/Users/ciphe/AppData/Local/Temp/flutter_tools.5a76b99f/flutter_test_listener.48a70eb7/listener.dart 22:3  _testMain


To run this test again: C:\dev\tooling\flutter\bin\cache\dart-sdk\bin\dart.exe test C:/dev/projects/OSS/diaryvault/integration_test/test_guest_login.dart -p vm --plain-name "loading C:/dev/projects/OSS/diaryvault/integration_test/test_guest_login.dart"
02:17 +0 -1: Some tests failed.

```

Comment (PratyushChauhan) [reply integration_test/test_guest_login.dart:9]: updated function
```
Future<void> main() async{
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
  await di.init();


  testWidgets('Guest login test, click on sign up button to move reveal guest login button'
      'click the guest login button to enter HomePage', (WidgetTester tester) async {
    await tester.pumpWidget(App());
    await tester.pumpAndSettle();
    await tester.tap(find.text('Sign up'));
    await tester.pumpAndSettle();
    await tester.tap(find.text('Continue as guest'));
    await tester.pumpAndSettle();
    expect(find.byType(HomePage), findsOneWidget);
    expect(find.byType(AuthPage), findsNothing);
  });
}
```

Comment (SankethBK) [reply integration_test/test_guest_login.dart:9]: Hey sorry for late reply, try calling `di.init` in `setUpAll` function. ref: https://stackoverflow.com/questions/62548585/how-to-use-get-it-flutter-dependency-injection-in-the-widget-test

**Comment by SankethBK** [comment]
This works, you can use this starting point

```
import 'package:dairy_app/app/view/app.dart';
import 'package:dairy_app/core/pages/home_page.dart';
import 'package:dairy_app/features/auth/presentation/pages/auth_page.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';
import 'package:dairy_app/core/dependency_injection/injection_container.dart'
    as di;

Future<void> main() async {
  setUpAll(() async {
    IntegrationTestWidgetsFlutterBinding.ensureInitialized();
    await dotenv.load();
    await di.init();
  });

  testWidgets(
      'Guest login test, click on sign up button to move reveal guest login button'
      'click the guest login button to enter HomePage',
      (WidgetTester tester) async {
    await tester.pumpWidget(const App());
    await tester.pumpAndSettle();
    await tester.tap(find.text('Continue as guest'));
    await tester.pumpAndSettle();
    expect(find.byType(HomePage), findsOneWidget);
  });
}

```

you can click on FloatingActionButton, create a note with some title and body and check if it appears in home page after saving

**Comment by SankethBK** [comment]
> This works for login but we get the following error if we try to make a note and save it in guest mode. image

Can you push the code, I'll try running locally

**Comment by PratyushChauhan** [comment]
On it. One more this, these logs were shown after I manually tried to create the note since the test kept failing.

**Comment by SankethBK** [comment]
Hi, this test will pass, i think the issue occurs when we try to add text into the rich text editor. So let's add only title for now 

```import 'package:dairy_app/app/view/app.dart';
import 'package:dairy_app/core/pages/home_page.dart';
import 'package:dairy_app/core/widgets/glassmorphism_cover.dart';
import 'package:dairy_app/features/notes/presentation/pages/note_create_page.dart';
import 'package:dairy_app/features/notes/presentation/widgets/note_save_button.dart';
import 'package:dairy_app/features/notes/presentation/widgets/note_title_input_field.dart';
import 'package:dairy_app/features/notes/presentation/widgets/notes_close_button.dart';
import 'package:dairy_app/features/notes/presentation/widgets/rich_text_editor.dart';
import 'package:flutter/material.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:flutter_quill/flutter_quill.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';
import 'package:dairy_app/core/dependency_injection/injection_container.dart'
    as di;

Future<void> main() async {
  setUpAll(() async {
    IntegrationTestWidgetsFlutterBinding.ensureInitialized();
    await dotenv.load();
    await di.init();
  });

  testWidgets('...', (WidgetTester tester) async {
    await tester.pumpWidget(const App());
    await tester.pumpAndSettle();
    await tester.tap(find.text('Continue as guest'));
    await tester.pumpAndSettle();
    expect(find.byType(HomePage), findsOneWidget);
    await tester.pumpAndSettle();
    await tester.tap(find.byType(FloatingActionButton));
    await tester.pumpAndSettle();
    // input some text in title field
    await tester.enterText(find.byType(NoteTitleInputField), 'Test Title');
    await tester.pumpAndSettle();

    expect(find.byType(QuillEditor), findsOneWidget);

    expect(
        find.descendant(
          of: find.byType(QuillEditor),
          matching: find.byType(TextFieldTapRegion),
        ),
        findsOneWidget);

    await tester.tap(find.byType(NoteSaveButton));
    await tester.pumpAndSettle();

    await Future.delayed(const Duration(seconds: 5));
    await tester.pumpAndSettle();
    expect(find.text('Test Title'), findsOneWidget);
    expect(find.byType(HomePage), findsOneWidget);
    expect(find.byType(NoteCreatePage), findsNothing);
  });
}
```
Also there was a bug during saving the note I have fixed it in latest master commit

**Comment by PratyushChauhan** [comment]
Makes sense, since we have to format strings before we pass them into the rich text editor. Thanks for your guidance.

**Comment by PratyushChauhan** [comment]
Will work on updation and deletion next.

**Comment by SankethBK** [comment]
> Makes sense, since we have to format strings before we pass them into the rich text editor.

Actually we should be able to enter raw text into the rich text editor since QuillEditor has `TextFieldTapRegion` as one of its child. It should handle the conversion to json internally. Not sure why its not working. Anyway we can consider title only for this issue.

**Comment by SankethBK** [comment]
Found some reference for how they are doing it in widget tests of `flutter_quill`. It looks a bit complicated, since we don't have access to `RawEditor` outside of the package. So its better to ignore this for now https://github.com/SankethBK/dairy_app/blob/f62ea47d797507477ef412c93232e17e0f02b6c7/packages/flutter_quill/lib/src/test/widget_tester_extension.dart

**Comment by PratyushChauhan** [comment]
Alright

**Comment by PratyushChauhan** [comment]
Add note updation test.

**Comment by PratyushChauhan** [comment]
Add integration test for note deletion.
Please review and let me know what other tests I should add.

**Comment by SankethBK** [comment]
> Add integration test for note deletion. Please review and let me know what other tests I should add.

Thank you, will review it by today

**Comment by SankethBK** [comment]
Let's have all `testWidgets` in a single file `note_crud_tests.dart` and create a group like this

```

void main() {
  group('MyWidget Tests', () {
    testWidgets('Test Case 1', (WidgetTester tester) async {
      // Your test case code here
    });

    testWidgets('Test Case 2', (WidgetTester tester) async {
      // Your test case code here
    });
    
}

```

It would be easier to segregate if we add more tests in future

**Comment by PratyushChauhan** [comment]
@SankethBK should I make helper functions for widgets tests and store them somewhere else for better readability?

eg:

```
void main() {
  group('MyWidget Tests', () {
    testWidgets('Test Case 1', (WidgetTester tester) async {
      await testTestCase1(tester);
    });

    testWidgets('Test Case 2', (WidgetTester tester) async {
      await testTestCase2(tester);
    });
  });
}

```

**Comment by SankethBK** [comment]
> @SankethBK should I make helper functions for widgets tests and store them somewhere else for better readability?
> 
> eg:
> 
> ```
> void main() {
>   group('MyWidget Tests', () {
>     testWidgets('Test Case 1', (WidgetTester tester) async {
>       await testTestCase1(tester);
>     });
> 
>     testWidgets('Test Case 2', (WidgetTester tester) async {
>       await testTestCase2(tester);
>     });
>   });
> }
> ```

@PratyushChauhan not required, we can do that in future if all tests have lot of repeated code

**Comment by PratyushChauhan** [comment]
@SankethBK  tried to merge tests into one file, encountered some problems which i think might be relating to dependecy initialization. Pushed changes for your reference.

Logs:
image

**Comment by SankethBK** [comment]
Second test is starting right where first test was stopped, so its failing. Because its trying to initialize everything again, but they are already initalized

**Comment by SankethBK** [comment]
Let's have them in different files only for now, like it was earlier.

**Comment by PratyushChauhan** [comment]
pushed refactorings.

**Comment by SankethBK** [comment]
Looks good, there is some linter error related to formatting, can you run this command
```
dart format --line-length 80 .
```
It will format the files, you can push those changes

**Comment by PratyushChauhan** [comment]
got it

**Comment by PratyushChauhan** [comment]
checks completed

**Review by SankethBK — APPROVED** [review]
Thanks for contributing!

**Comment by PratyushChauhan** [comment]
Always a pleasure.

**Pull request merged**